pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,508 @@
+	waitNext       map[string]time.Time  // sanction connections to a peer, key is overlay string and value is time to next retry
+	waitNextMu     sync.Mutex            // synchronize map
+	logger         logging.Logger        // logger
+	quit           chan struct{}         // quit channel
+	done           chan struct{}         // signal that `manage` has quit
+}
+
+// New returns a new Kademlia.
+func New(o Options) *Kad {
+	if o.SaturationFunc == nil {
+		o.SaturationFunc = binSaturated
+	}
+
+	k := &Kad{
+		base:           o.Base,
+		discovery:      o.Discovery,
+		addressBook:    o.AddressBook,
+		p2p:            o.P2P,
+		saturationFunc: o.SaturationFunc,
+		connectedPeers: pslice.New(maxBins),
+		knownPeers:     pslice.New(maxBins),
+		manageC:        make(chan struct{}, 1),
+		waitNext:       make(map[string]time.Time),
+		logger:         o.Logger,
+		quit:           make(chan struct{}),
+		done:           make(chan struct{}),
+	}
+
+	go k.manage()
+	return k
+}
+
+// manage is a forever loop that manages the connection to new peers
+// once they get added or once others leave.
+func (k *Kad) manage() {
+	var peerToRemove swarm.Address
+	defer close(k.done)
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-k.quit
+		cancel()
+	}()
+
+	for {
+		select {
+		case <-k.quit:
+			return
+		case <-k.manageC:
+			err := k.knownPeers.EachBinRev(func(peer swarm.Address, po uint8) (bool, bool, error) {
+
+				if k.connectedPeers.Exists(peer) {
+					return false, false, nil
+				}
+
+				k.waitNextMu.Lock()
+				if next, ok := k.waitNext[peer.String()]; ok && time.Now().Before(next) {
+					k.waitNextMu.Unlock()
+					return false, false, nil
+				}
+				k.waitNextMu.Unlock()
+
+				currentDepth := k.NeighborhoodDepth()
+				if k.saturationFunc(po, currentDepth, k.connectedPeers) {
+					return false, true, nil // bin is saturated, skip to next bin
+				}
+
+				ma, err := k.addressBook.Get(peer)
+				if err != nil {
+					// either a peer is not known in the address book, in which case it
+					// should be removed, or that some severe I/O problem is at hand
+
+					if errors.Is(err, storage.ErrNotFound) {
+						k.logger.Errorf("failed to get address book entry for peer: %s", peer.String())
+						peerToRemove = peer
+						return false, false, errMissingAddressBookEntry
+					}
+
+					return false, false, err
+				}
+				k.logger.Debugf("kademlia dialing to peer %s", peer.String())
+
+				err = k.connect(ctx, peer, ma, po)
+				if err != nil {
+					k.logger.Errorf("error connecting to peer %s: %v", peer, err)
+					// continue to next
+					return false, false, nil
+				}
+
+				k.waitNextMu.Lock()
+				delete(k.waitNext, peer.String())
+				k.waitNextMu.Unlock()
+
+				k.depthMu.Lock()
+				k.depth = k.recalcDepth()
+				k.depthMu.Unlock()
+
+				k.logger.Debugf("connected to peer: %s old depth: %d new depth: %d", peer, currentDepth, k.NeighborhoodDepth())
+
+				select {
+				case <-k.quit:
+					// shutting down
```

Comment:
🌷 maybe no need for this comment, it's kind of obvious as the code itself is nice & readable, imo.